agcaballero/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -57,106 +96,186 @@ boolean isReplaceableBy(Availability availability) {
    */
-  private List<TimeRange> availableTimeRanges (Availability[] minuteAvailabilities, 
-      Set<Availability> matchingAvailabilities, long requestDuration) {
-    // declare all variables to keep track of time ranges
-    int start = 0; 
-    ArrayList<TimeRange> times = new ArrayList<TimeRange>();
-    boolean thisMinuteAvailable = matchingAvailabilities.contains(minuteAvailabilities[start]);
-    // for first minute, this will have same value as thisMinuteAvailable but shouldn't affect result
-    boolean wasLastMinuteAvailable = thisMinuteAvailable; 
-
-    // go through minute availabilities array and create time ranges for all
-    // spots that match one of the availabilities in the matching availabilities set
-    for (int i = 0; i < minuteAvailabilities.length; i++) {
-      // gets availability of current minute (considered available if status matches one within set)
-      thisMinuteAvailable = matchingAvailabilities.contains(minuteAvailabilities[i]);
+  private Collection<TimeRange> findTimesBasedOnOptionalAttendeeAvailability(MeetingRequest request, 
+    Availability[] minuteAvailability, int maxOptionalAttendeesUnavailable) {
+
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    int start = 0;
+    boolean wasLastMinuteAvailable = minuteAvailability[0].areMandatoryAttendeesAllAvailable 
+        && minuteAvailability[0].numberOfOptionalAttendeesUnavailable <= maxOptionalAttendeesUnavailable;
+  
+    for (int i = 0; i < minuteAvailability.length; i++) {
+    
+      // checks if the current minute is available by checking that mandatory attendees are available
+      // and that the number of optional attendees that can't make it is under the max
+      boolean currentMinuteAvailable = minuteAvailability[i].areMandatoryAttendeesAllAvailable 
+        && minuteAvailability[i].numberOfOptionalAttendeesUnavailable <= maxOptionalAttendeesUnavailable;
 
       if (wasLastMinuteAvailable) {
         // If the previous minute was available, but the current minute is unavailable or if it's
         // the end of the day, this is the end of an available time range. If the time range is longer 
         // than the required duration, it's recorded as an available time range.
-        if (!thisMinuteAvailable  || i == minuteAvailabilities.length - 1) {
+        if (! currentMinuteAvailable || i == minuteAvailability.length - 1) {
           int end = i;
           int duration = end - start;
 
-          if (duration >= requestDuration) {
-            times.add(TimeRange.fromStartEnd(start, end - 1, true)); // add time range (inclusive of start & end) 
-          }
-          
-          // avoid an inconsistency b/c in case availableMinutes[availableMinutes.length - 1] is true,
-          // you shouldn't set this to false (& that would happen without this if)
-          if (!thisMinuteAvailable) {
-            wasLastMinuteAvailable = false;
+          if (duration >= request.getDuration()) {
```

Comment:
Nit: consider caching getDuration() as a local variable outside of the loop to avoid a method call on each iteration.

Reply:
I will do this. Thanks!

Reply:
Done